### DISCUSSION TITLE
Does RuboCop "check the AST" on the changes it is doing?

### DISCUSSION THREAD

**Discussion by stdedos** [rubocop/rubocop#13546] [Q&A] [answered]
Inspired by https://github.com/rubocop/rubocop/pull/13499#discussion_r1858956541, and what I know by https://black.readthedocs.io/en/stable/*:

Could RuboCop compare the AST, to avoid creating issues such as when operator precedence is involved, or (implict) block starting which does not require parentheses, but the rewrite does require parentheses to keep semantics^? 

I feel like, if that were the case, issues like https://github.com/rubocop/rubocop/issues/13299 would not have happened.

---

*: Ik that black is only a formatter, therefore "maybe" easier than RuboCop. But the `{ |x| }` vs `do ... end` is also one of _style_, I'd say in this context.
^: or whichever is the correct Ruby terminology 😅

**Comment by dvandersluis** [comment, marked as answer]
I was actually looking into doing that after I made that comment but there are reasons that the AST could change without changing operator precedence. 

For instance, this test: https://github.com/rubocop/rubocop/blob/1b237c643c9583846b4599b4c5ba32411a29091d/spec/rubocop/cop/style/block_delimiters_spec.rb#L407-L426

The AST changes (a `begin` node is inserted), but precedence is not affected and the code behaviour doesn't change.

Adding parentheses around a node also inserts a `begin` node, BTW.

**Reply by stdedos** [reply]
On _very wrong terms_, replacing `do` with `{` "removes an implicit begin node", which the `begin` then re-adds explicitly.

In that prism, it looks like a net-0 change.

**Reply by dvandersluis** [reply]
Sure, but that still means that the AST will have changed.
